### PR TITLE
[proof] Miscellaneous refactorings

### PR DESCRIPTION
**Pull request by ejgallego**
This is a continuation of #11016 , with several minor refactorings.

Overlays:
- https://github.com/mattam82/Coq-Equations/pull/264
- https://github.com/mit-plv/rewriter/pull/7
- https://github.com/LPCIC/coq-elpi/pull/125

### CODE REVIEW

**Comment by ejgallego** [comment]
Ok, this should be ready now; mostly a preparation for upcoming PRs.

**Comment by ejgallego** [comment]
Anyone willing to assign, maybe @SkySkimmer ?

File: pretyping/pretyping.mli [hunk 1]
```diff
@@ -44,7 +44,8 @@ type typing_constraint =
   | OfType of types (** A term of the expected type *)
   | WithoutTypeConstraint (** A term of unknown expected type *)
 
-type inference_hook = env -> evar_map -> Evar.t -> evar_map * constr
+(* Some = did suceeed *)
```
Comment (gares) [inline pretyping/pretyping.mli]: this comment is both useless and mistyped

**Review by gares — APPROVED** [review]
files I own are ok

File: pretyping/pretyping.mli (hunk 1 above)
Comment (ejgallego) [reply pretyping/pretyping.mli]: Thanks for the review @gares ; comment amended, I hope it is better now.

File: tactics/declare.ml [hunk 2]
```diff
@@ -351,10 +351,10 @@ let declare_private_constant ?role ?(local = ImportDefaultBehavior) ~name ~kind
   let eff = { Evd.seff_private = eff; Evd.seff_roles; } in
   kn, eff
 
-let inline_private_constants ~univs env ce =
+let inline_private_constants ~uctx env ce =
   let body, eff = Future.force ce.proof_entry_body in
   let cb, ctx = Safe_typing.inline_private_constants env (body, eff.Evd.seff_private) in
-  let univs = UState.merge ~sideff:true Evd.univ_rigid univs ctx in
+  let univs = UState.merge ~sideff:true Evd.univ_rigid uctx ctx in
```
Comment (SkySkimmer) [inline tactics/declare.ml]:
```suggestion
  let uctx = UState.merge ~sideff:true Evd.univ_rigid uctx ctx in
```

File: tactics/pfedit.mli [hunk 3]
```diff
@@ -65,7 +65,7 @@ val build_constant_by_tactic
   :  name:Id.t
   -> ?opaque:Proof_global.opacity_flag
   -> uctx:UState.t
-  -> named_context_val
+  -> sign:named_context_val
```
Comment (SkySkimmer) [inline tactics/pfedit.mli]: Not sure about the value of this change but won't oppose

File: vernac/declareObl.ml [hunk 4]
```diff
@@ -286,7 +288,7 @@ let progmap_add n prg =
 let progmap_replace prg' =
   Lib.add_anonymous_leaf (input (map_replace prg'.prg_name prg' !from_prg))
 
-let obligations_solved prg = Int.equal (snd prg.prg_obligations) 0
+let obligations_solved prg = Int.equal (prg.prg_obligations.remaining) 0
```
Comment (SkySkimmer) [inline vernac/declareObl.ml]:
```suggestion
let obligations_solved prg = Int.equal prg.prg_obligations.remaining 0
```

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Will leave a couple days in case obligations experts want to comment on the last few commits (@mattam82?)

File: doc/plugin_tutorial/tuto1/src/simple_declare.ml [hunk 5]
```diff
@@ -1,10 +1,10 @@
-let edeclare ?hook ~name ~poly ~scope ~kind ~opaque sigma udecl body tyopt imps =
+let edeclare ?hook ~name ~poly ~scope ~kind ~opaque sigma udecl body tyopt impargs =
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply doc/plugin_tutorial/tuto1/src/simple_declare.ml]: Fixed thanks.

File: tactics/pfedit.ml [hunk 6]
```diff
@@ -129,18 +129,18 @@ let build_constant_by_tactic ~name ?(opaque=Proof_global.Transparent) ctx sign ~
...
+  let ce, status, univs = build_constant_by_tactic ~name uctx sign ~poly typ tac in
   let cb, univs =
     if side_eff then Declare.inline_private_constants ~univs env ce
     else
       (* GG: side effects won't get reset: no need to treat their universes specially *)
       let (cb, ctx), _eff = Future.force ce.Declare.proof_entry_body in
       cb, UState.merge ~sideff:false Evd.univ_rigid univs ctx
   in
-  cb, status, univs
+  cb, ce.Declare.proof_entry_type, status, univs
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply tactics/pfedit.ml]: `typ` is `EConstr` we return a `Constr.t`

File: tactics/declare.ml (hunk 2 above)
Comment (ejgallego) [reply tactics/declare.ml]: Fixed, also clarified a couple other variables.

File: tactics/pfedit.mli (hunk 3 above)
Comment (ejgallego) [reply tactics/pfedit.mli]: It was used in a few other places, it was just to label all the "accesory" args here.

File: vernac/declareObl.ml (hunk 4 above)
Comment (ejgallego) [reply vernac/declareObl.ml]: Done thanks.

**Comment by SkySkimmer** [comment]
A merge conflict appeared.

**Comment by ejgallego** [comment]
> A merge conflict appeared.

Fixed; also fixed overlay being out of date.

**Comment by ejgallego** [comment]
Failure on win was spurious.

**Comment by ejgallego** [comment]
Thanks!

**Pull request merged**